### ISSUE TITLE
jenkins_plugin code broken due to new -remoting cli change.

### ISSUE THREAD

**Issue by stefhen** [chef-cookbooks/jenkins#602]
### Cookbook version

`5.0.0`

### Chef-client version

`Chef: 12.19.36`

### Platform Details

```bash
root@ip-192-168-0-228:/home/rightscale/.chef# lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.1 LTS
Release:        16.04
Codename:       xenial
```
Running in AWS.

### Scenario:

Plugin installation fails, appears to be due to the cli update that isn't based on remoting, announced today here: https://jenkins.io/blog/2017/04/26/security-advisory/

```bash
---- Begin output of "/usr/lib/jvm/java-8-openjdk-amd64/bin/java" -jar "/home/rightscale/.chef/local-mode-cache/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /home/rightscale/.chef/local-mode-cache/cache/swarm-3.4.plugin -name swarm  ----
STDOUT: /home/rightscale/.chef/local-mode-cache/cache/swarm-3.4.plugin is neither a valid file, URL, nor a plugin artifact name in the update center
STDERR: ERROR: Error occurred, see previous output.
```
### Steps to Reproduce:

Run `jenkins_plugin` code and get the error above.

Example:

```ruby
jenkins_plugin 'swarm' do
  version node['rsc_jenkins']['swarm']['version']
  install_deps true
  notifies :restart, 'service[jenkins]', :immediately
end
```
### Expected Result:

Plugin should be installed.

### Actual Result:

Plugin install fails with error above instead of being installed.

It appears the `-remoting` option needs to be passed to the cli options in `jenkins_plugin` for this to now work:

```bash
 java -jar /home/rightscale/.chef/local-mode-cache/cache/jenkins-cli.jar -remoting -s http://localhost:8080  install-plugin /home/rightscale/.chef/local-mode-cache/cache/swarm-3.4.plugin -name swarm
Installing a plugin from local file: /home/rightscale/.chef/local-mode-cache/cache/swarm-3.4.plugin
```

**Comment by bflad** [comment]
Also seeing this issue.

**Comment by daften** [comment]
Related to #599

**Comment by kbaltrinic** [comment]
+1 Major issue for us.

**Comment by Poohblah** [comment]
Does anybody know of a workaround for this issue?

I've tried downgrading the jenkins cookbook to < 5.0.0, which doesn't seem to have any effect, and I've tried the following as suggested in the OP:
```
jenkins_plugin 'plugin_name' do
  options '-remoting'
end
```
which just gives me:
```
================================================================================
Error executing action `install` on resource 'jenkins_plugin[ansicolor]'
================================================================================

Mixlib::ShellOut::ShellCommandFailed
------------------------------------
Expected process to exit with [0], but received '2'
---- Begin output of "/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /tmp/kitchen/cache/ansicolor-0.5.0.plugin -name ansicolor -remoting ----
STDOUT: 
STDERR: ERROR: "-remoting" is not a valid option
java -jar jenkins-cli.jar install-plugin SOURCE ... [-deploy] [-name VAL] [-restart]
Installs a plugin either from a file, an URL, or from update center. 
 SOURCE    : If this points to a local file (‘-remoting’ mode only), that file
      will be installed. If this is an URL, Jenkins downloads the URL
      and installs that as a plugin. If it is the string ‘=’, the file
      will be read from standard input of the command, and ‘-name’ must
      be specified. Otherwise the name is assumed to be the short name
      of the plugin in the existing update center (like ‘findbugs’), and
      the plugin will be installed from the update center.
 -deploy   : Deploy plugins right away without postponing them until the reboot.
 -name VAL : If specified, the plugin will be installed as this short name
      (whereas normally the name is inferred from the source name
      automatically).
 -restart  : Restart Jenkins upon successful installation.
---- End output of "/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /tmp/kitchen/cache/ansicolor-0.5.0.plugin -name ansicolor -remoting ----
Ran "/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /tmp/kitchen/cache/ansicolor-0.5.0.plugin -name ansicolor -remoting returned 2

Cookbook Trace:
---------------
/tmp/kitchen/cache/cookbooks/jenkins/libraries/_executor.rb:85:in `execute!'
/tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:317:in `install_plugin_from_url'
/tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:289:in `install_plugin_from_update_center'
/tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:124:in `block (2 levels) in <class:JenkinsPlugin>'
/tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:154:in `block in <class:JenkinsPlugin>'

Resource Declaration:
---------------------
# In /tmp/kitchen/cache/cookbooks/cub_jenkins/recipes/default.rb

 61:   jenkins_plugin(p) { options '-remoting' }
 62: end

Compiled Resource:
------------------
# Declared in /tmp/kitchen/cache/cookbooks/cub_jenkins/recipes/default.rb:61:in `block in from_file'

jenkins_plugin("ansicolor") do
  action [:install]
  retries 0
  retry_delay 2
  default_guard_interpreter :default
  declared_type :jenkins_plugin
  cookbook_name "cub_jenkins"
  recipe_name "default"
  options "-remoting"
end

Platform:
---------
x86_64-linux
```
Platform is RHEL 7.3.

**Comment by charlottestjohn** [comment]
Blocked on upgrading past Jenkins 2.53 based off of this issue.

**Comment by daften** [comment]
@Poohblah  Look at #603 for a fix

**Comment by iennae** [comment]
Thanks for the PR folks. Please check to see if cookbook version 5.0.1 resolves this issue for you.

**Comment by oberones** [comment]
5.0.1 actually causes this issue for me. I've had to pin my cookbooks to 5.0.0 for now.

**Comment by daften** [comment]
@oberones What version of jenkins are you on?

**Comment by oberones** [comment]
@daften I'm using 2.32.1, which does not have the remoting option available in the cli.

**Comment by daften** [comment]
@oberones have you tried putting executor['protocol'] to nil by overriding it in a wrapper cookbook, environment, ...? That should disable the -remoting option

**Comment by Poohblah** [comment]
5.0.1 seems to fix the issues with `jenkins_plugin` for me.

**Comment by oberones** [comment]
@daften I haven't tried that but I bet it will work.

**Comment by isuftin** [comment]
Version 5.0.1 does not fix this for me. I am unable to use the SSH protocol because Jenkins by default starts with SSH disabled so the SSH protocol causes errors when attempting to do anything.

I am trying to use the http protocol as the remoting protocol is deprecated and causes warnings. 

Using the http protocol causes the same errors.

I've attempted to set the protocol to nil. This gives me:

```
           ================================================================================
           Error executing action `install` on resource 'jenkins_plugin[display-url-api]'
           ================================================================================

           Mixlib::ShellOut::ShellCommandFailed
           ------------------------------------
           Expected process to exit with [0], but received '5'
           ---- Begin output of "/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /tmp/kitchen/cache/display-url-api-2.0.plugin -name display-url-api  ----
           STDOUT: /tmp/kitchen/cache/display-url-api-2.0.plugin is neither a valid file, URL, nor a plugin artifact name in the update center
           STDERR: ERROR: Error occurred, see previous output.
           May 11, 2017 3:57:16 PM hudson.cli.CLI$5 run
           WARNING: null
           java.io.IOException: Stream is closed
                at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.checkError(HttpURLConnection.java:3512)
                at sun.net.www.protocol.http.HttpURLConnection$StreamingOutputStream.write(HttpURLConnection.java:3486)
                at java.io.DataOutputStream.writeInt(DataOutputStream.java:197)
                at hudson.cli.PlainCLIProtocol$EitherSide.send(PlainCLIProtocol.java:175)
                at hudson.cli.PlainCLIProtocol$ClientSide.sendEndStdin(PlainCLIProtocol.java:347)
                at hudson.cli.CLI$5.run(CLI.java:671)
           ---- End output of "/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /tmp/kitchen/cache/display-url-api-2.0.plugin -name display-url-api  ----
           Ran "/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 install-plugin /tmp/kitchen/cache/display-url-api-2.0.plugin -name display-url-api  returned 5

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/jenkins/libraries/_executor.rb:87:in `execute!'
           /tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:317:in `install_plugin_from_url'
           /tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:289:in `install_plugin_from_update_center'
           /tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:124:in `block (2 levels) in <class:JenkinsPlugin>'
           /tmp/kitchen/cache/cookbooks/jenkins/libraries/plugin.rb:154:in `block in <class:JenkinsPlugin>'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/owi_jenkins/recipes/plugins.rb

            17:   jenkins_plugin plugin do
            18:     version version if version
            19:     notifies :restart, 'service[jenkins]', :delayed
            20:   end
            21: end
```

**Comment by isuftin** [comment]
Note that manually adding `file://` works in the call to the CLI...

`"/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 -"http" install-plugin file:///tmp/kitchen/cache/display-url-api-2.0.plugin -name display-url-api
Installing a plugin from file:/tmp/kitchen/cache/display-url-api-2.0.plugin`

However, that's not something that I can fix in my wrapper cookbook. That needs a fix right around here: https://github.com/chef-cookbooks/jenkins/blob/master/libraries/plugin.rb#L309

**Comment by isuftin** [comment]
If I manually add `file://` to the path when logged into the VM: `"/usr/lib/jvm/java/bin/java" -jar "/tmp/kitchen/cache/jenkins-cli.jar" -s http://localhost:8080 -"http" install-plugin file:///tmp/kitchen/cache/display-url-api-2.0.plugin -name display-url-api`

... this works fine. so it seems like Jenkins does not see a direct filepath without `file://` as a valid URI

**Comment by iennae** [comment]
@stefhen can you update the title of this issue to reflect the problem that you are seeing rather than ascribing to remoting option being added? I think there are some conflicting reports being added to this and it's hard to track down what is broken/fixed for folks. I think the file:// issue that was pulled in today will fix your issue but I don't want to assume. Based on the change, then I can start asking folks who still have problems to separate out their issues rather than trying to solve everything about plugin installation in this issue (because plugin installation is a big problem that isn't going to be solved in one big change). Thanks!

**Comment by stefhen** [comment]
Hi @iennae This was actually fixed in `5.0.1` and can be closed. https://github.com/chef-cookbooks/jenkins/blob/v5.0.1/CHANGELOG.md#501-2017-05-01

**Comment by iennae** [comment]
Thanks for updating @stefhen. For folks still having issues, please verify that a different issue hasn't already been created. If not, create a new issue so that we can make sure to track the specific changes required. Plugin installation is complex and we are working to do a rehaul on this completely.

**Comment by tknerr** [comment]
Even with `-remoting` being the explicit default now, it doesn't work with Jenkins 2.60.1 and newer, because the remoting CLI is disabled by default, and needs to be re-enabled.

See my comment here:
https://github.com/chef-cookbooks/jenkins/pull/603#issuecomment-317728006
